main/tests/server/src/com/google/refine/expr/functions/arrays/JoinTests.java

```diff
@@ -28,13 +28,35 @@
 
 import org.testng.annotations.Test;
 
+import com.google.refine.RefineTest;
+import com.google.refine.expr.ParsingException;
 import com.google.refine.util.TestUtils;
 
-public class JoinTests {
+public class JoinTests extends RefineTest {
     @Test
     public void serializeJoin() {
         String json = "{\"description\":\"Returns the string obtained by joining the array a with the separator sep\",\"params\":\"array a, string sep\",\"returns\":\"string\"}";
         TestUtils.isSerializedTo(new Join(), json);
     }
+
+    @Test
+    public void joinArray() throws ParsingException {
+        String[] test = {"[2,1,3].join('|')", "2|1|3"};
+        parseEval(bindings, test);
+
+        // TODO: This is current behavior, but is it what we want?
+        String[] test1 = {"[2,null,3].join(', ')", "2, 3"};
+        parseEval(bindings, test1);
+
+        String[] test2 = {"['z','b','c','a'].join('-')", "z-b-c-a"};
+        parseEval(bindings, test2);
+
+        // TODO: Do we really want the following two tests to return different results?
+        String[] test3 = {"['z', null,'c','a'].join('-')", "z-c-a"};
+        parseEval(bindings, test3);
+        String[] test4 = {"['z', '','c','a'].join('-')", "z--c-a"};
```

Comment:
Yes, they should return differently.  I think that was agreed in #820 and it's linked issues/pr's
We might even consider a combined test for null/empty?
`String[] test4 = {"['z', '', null,'c','a'].join('-')", "z---c-a"};`

Reply:
As I had mentioned before, other data tools have Null morpheme as a preference for the user for their data.  We might eventually want to do that.  I do feel somewhat disappointed that the whole thing felt rushed, and we really didn't get more opinions come in from the community on this...some just expected things to work well for null/empty strings no matter.  Others expected to have more control given to them and not less taken away.  We seemed to have taken some control away and saying "We shall always treat them differently now, no matter what, because we're too lazy to change all the places in our code to deal with Null morpheme".
Maybe we can make that a goal for 2021/22 ... add a preference for null/empty string treated as equivalent?  But it feels backwards to even say that.

Reply:
Hmm, Java8 has Optional? https://stackoverflow.com/questions/21936503/get-empty-string-when-null

Reply:
I'm afraid despite lots of effort and discussion we still don't have consistency in situations where a `null` is forced to a string. I think it's definitely worth another review and attempt to get to a reasonable approach but as a separate issue to this. So pending further discussion & agreement I feel its safer to keep the current behaviour for `join`

Reply:
@thadguidry Your proposed test would fail. The result would actually be `z--c-a` which is one of the weirdly inconsistent things about this.

For the record, I wasn't actually considering any unrelated changes to `join()` as part of this PR. I was just flagging the inconsistency for future review.

I added the tests because I was working in the neighborhood of completely untested functions. I added the comments when I had to change the tests to match the inconsistent behavior of our current implementation.

Reply:
@tfmorris Thanks Tom.  I see on other issue comments that you are diving into it all.  We'll let you work it all out :-)  Thanks so much!